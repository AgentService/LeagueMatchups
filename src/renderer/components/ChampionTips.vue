<template>
		<div class="card-header">
			General Tips <!-- Header title -->
		</div>
		<transition name="fade" mode="out-in">
			<div class="overflow-auto d-flex  flex-grow-1" v-if="championTips && Object.keys(championTips).length > 0">
				<div class="card-body tips-card" :key="selectedTip">
					<div v-if="!selectedTip" class="flex-container d-flex ">
						<!-- Iterate over each key in championTips -->
						<div v-for="(value, key) in championTips" :key="key" class="flex-item">
							<!-- Button for each tip -->
							<button type="button"
								class="btn text-secondary justify-content-center align-items-center"
								@click="selectedTip = key">
								<i :class="['me-2', getIconForKey(key)]" aria-hidden="true"></i>
								<div class="text-content">{{ formatCategory(key) }}</div>
							</button>
						</div>
					</div>
					<div v-else class="content-container">
						<div class="details-header d-flex align-items-center">
							<!-- Back Button -->
							<button type="button" class="btn" @click="selectedTip = null">
								<i class="fas fa-chevron-left"></i>
							</button>
							<!-- Title next to back button -->
							<h6 class="details-title">{{ formatCategory(selectedTip) }}</h6>
						</div>
						<div class="tip-details">
							<!-- Debugging-Log -->
							<p>{{ championTips[selectedTip]?.long }}</p>
						</div>
					</div>
				</div>
			</div>
			<div v-else class="card-body tips-card">
				<div class="content-container d-flex align-items-center justify-content-center">
					<div class="no-tips-message text-center">
						<p>No tips available for this champion.</p>
					</div>
				</div>
			</div>
		</transition>
</template>


<script setup>
import { computed, ref, defineProps, watch } from 'vue';
import { useStore } from 'vuex';

const props = defineProps({
	champion: Object,
	instanceId: Number
});

const store = useStore();
const championId = ref(props.champion?.id);
const championTips = computed(() => {
	return store.getters['champions/getChampionTips'](championId);
});
const selectedTip = ref(null);

const formatCategory = (key) => {
	if (!key) return '';
	key = key.replace(/([A-Z])/g, ' $1').toLowerCase();
	key = key.replace(/^\w|\s\w/g, (str) => str.toUpperCase());
	return key.trim();
};

const getIconForKey = (category) => {
	const icons = {
		'identity': 'fa-solid fa-user-circle',
		'strengths': 'fa-solid fa-thumbs-up',
		'weaknesses': 'fa-solid fa-thumbs-down',
		'earlygame': 'fa-solid fa-hourglass-start',
		'midgame': 'fa-solid fa-hourglass-half',
		'lategame': 'fa-solid fa-hourglass-end',
		'teamfight': 'fa-solid fa-shield-alt',
		'counter': 'fa-solid fa-ban',
		'mindset': 'fa-solid fa-brain',
		'insights': 'fa-solid fa-lightbulb',
		// Define more icons for other categories if needed
	};
	return icons[category] || 'fa-solid fa-question-circle'; // Fallback icon
};

watch(() => props.champion, (newChampion) => {
	if (newChampion && newChampion.id !== championId.value) {
		championId.value = newChampion.id;
	}
}, { immediate: true });

</script>


<style>
.tips-card {
	display: flex;
	min-height: 100%;
	flex-direction: column;
	padding: 1rem;
	box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);
	hyphens: auto; 
	text-align: justify;
}
.title-bar {
	text-align: center;
}



.flex-container {
	display: flex;
	flex-wrap: wrap;
	gap: 1rem;
	justify-content: start;
}

.flex-item {
	flex: 0 1 calc(50% - 10px); /* 2 items per row for smaller screens */
    display: flex;
    justify-content: center;
    align-items: center;
	
}

/* Enter and leave transitions */
.fade-enter-active,
.fade-leave-active {
	transition: opacity 0.10s;
}

.fade-enter-from,
.fade-leave-to {
	opacity: 0;
}

.grid-item {
	display: flex;
	flex-direction: column;
	height: 100%;

}

.tip-details {
	overflow: auto;
	flex-grow: 1;
	padding: 1rem;
	font-weight: 400;
}

.details-header {
	display: flex;
	align-items: center;
	flex-shrink: 0;
}

.details-title {
	flex-grow: 1;
	font-weight: 600;
	margin-bottom: .1rem;
}

.content-container {
	display: flex;
	flex-direction: column;
	flex-grow: 1;
	height: 100%;
}
</style>