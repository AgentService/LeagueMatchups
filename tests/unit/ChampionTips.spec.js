import { shallowMount } from "@vue/test-utils";
import ChampionTips from "@/renderer/components/ChampionTips.vue";
import { useStore } from "vuex";
import { createStore } from 'vuex';

jest.mock("vuex", () => ({
  useStore: jest.fn(() => ({
    state: {}, // Mock state
    getters: {
      "champions/getChampionTips": jest.fn().mockReturnValue({
        /* Mock return value */
        tipKey: { long: "Mocked tip content" },
      }),
    },
  })),
}));
it("computes champion tips correctly", async () => {
  const wrapper = shallowMount(ChampionTips, {
    props: {
      champion: { id: "champion1", name: "Champion One" },
      instanceId: 1,
    },
  });

  // Use the same formatCategory function to ensure we have the correct text to match
  const formattedText = wrapper.vm.formatCategory("tipKey");

  // Find the button by looking for .text-content div with the matching text
  const buttons = wrapper.findAll("button");
  let foundButton;
  for (const buttonWrapper of buttons) {
    const textContentDiv = buttonWrapper.find(".text-content");
    if (
      textContentDiv.exists() &&
      textContentDiv.text().includes(formattedText)
    ) {
      foundButton = buttonWrapper;
      break;
    }
  }

  if (foundButton) {
    await foundButton.trigger("click");
    await wrapper.vm.$nextTick();
    expect(wrapper.text()).toContain("Mocked tip content");
  } else {
    throw new Error(
      `Button with formatted text '${formattedText}' was not found`
    );
  }
});

describe("ChampionTips.vue", () => {
  it("renders correctly with mock store", () => {
    const wrapper = shallowMount(ChampionTips, {
      props: {
        champion: { id: "champion1", name: "Champion One" },
        instanceId: 1,
      },
    });

    expect(wrapper.exists()).toBe(true);
    expect(wrapper.find(".title-bar").text()).toContain("General Tips");
    // Other assertions to check if the component renders as expected
  });

  it("renders nothing when championTips is empty", () => {
	// Mock store to return an empty object for championTips
	useStore.mockReturnValue({
	  state: {},
	  getters: {
		"champions/getChampionTips": jest.fn().mockReturnValue({}),
	  },
	});
  
	const wrapper = shallowMount(ChampionTips, {
	  props: {
		champion: { id: "champion1", name: "Champion One" },
		instanceId: 1,
	  },
	});
  
	// Check if the component renders nothing
	expect(wrapper.html()).toBe('<!--v-if-->');
  
	// Ensure no errors are thrown and the component still exists
	expect(wrapper.exists()).toBe(true);
  });
  
  
  it("matches the snapshot", () => {
	const wrapper = shallowMount(ChampionTips, {
	  props: {
		champion: { id: "champion1", name: "Champion One" },
		instanceId: 1,
	  },
	});
	expect(wrapper.html()).toMatchSnapshot();
  });
  
  
});
  
it("renders nothing when championTips is empty", () => {
	// Mock useStore to return an empty object for championTips
	useStore.mockImplementation(() => ({
	  state: {
		champions: {
		  championTips: {}
		}
	  },
	  getters: {
		'champions/getChampionTips': () => () => ({})
	  }
	}));
  
	const wrapper = shallowMount(ChampionTips, {
	  props: {
		champion: { id: "champion1", name: "Champion One" },
		instanceId: 1,
	  },
	});
  
	// Check if the component renders nothing
	expect(wrapper.html()).toBe('<!--v-if-->');
  });

  
    it("renders correctly with different champion props", async () => {
	const wrapper = shallowMount(ChampionTips, {
	  props: {
		champion: { id: "champion2", name: "Champion Two" },
		instanceId: 2,
	  },
	});
	// Assertions to check if it renders correctly for a different champion
	expect(wrapper.vm.championId).toBe("champion2");
  });

  