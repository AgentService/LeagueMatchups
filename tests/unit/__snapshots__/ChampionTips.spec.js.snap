// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChampionTips.vue matches the snapshot 1`] = `"<!--v-if-->"`;
